src/catchup/ApplyBucketsWork.cpp

```diff
@@ -117,19 +126,28 @@ ApplyBucketsWork::onStart()
         CLOG(DEBUG, "History") << "ApplyBuckets : starting level[" << mLevel
                                << "].curr = " << i.curr;
         mApplying = true;
+        mBucketApplyStart.Mark();
     }
 }
 
 void
 ApplyBucketsWork::onRun()
 {
-    if (mSnapApplicator && *mSnapApplicator)
+    try
     {
-        mSnapApplicator->advance();
+        if (mSnapApplicator && *mSnapApplicator)
+        {
+            mSnapApplicator->advance();
+        }
+        else if (mCurrApplicator && *mCurrApplicator)
+        {
+            mCurrApplicator->advance();
+        }
     }
-    else if (mCurrApplicator && *mCurrApplicator)
+    catch (...)
     {
-        mCurrApplicator->advance();
+        mBucketApplyFailure.Mark();
```

Comment:
this probably belongs in the `OnFailure` family of handlers: from what I understand `Work` methods are not supposed to throw